zenith_utils/src/pq_proto.rs

```diff
@@ -0,0 +1,559 @@
+// Postgres protocol messages serialization-deserialization
+
+use anyhow::{anyhow, bail};
+use byteorder::{ReadBytesExt, WriteBytesExt, BE};
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+// use postgres_ffi::xlog_utils::TimestampTz;
+use std::collections::HashMap;
+use std::io;
+use std::io::{Read, Write};
+use std::str;
+
+pub type Oid = u32;
+pub type SystemId = u64;
+
+#[derive(Debug)]
+pub enum FeMessage {
+    StartupMessage(FeStartupMessage),
+    Query(FeQueryMessage), // Simple query
+    Parse(FeParseMessage), // Extended query protocol
+    Describe(FeDescribeMessage),
+    Bind(FeBindMessage),
+    Execute(FeExecuteMessage),
+    Close(FeCloseMessage),
+    Sync,
+    Terminate,
+    CopyData(Bytes),
+}
+
+#[derive(Debug)]
+pub struct FeStartupMessage {
+    pub version: u32,
+    pub kind: StartupRequestCode,
+    // optional params arriving in startup packet
+    pub params: HashMap<String, String>,
+}
+
+#[derive(Debug)]
+pub enum StartupRequestCode {
+    Cancel,
+    NegotiateSsl,
+    NegotiateGss,
+    Normal,
+}
+
+#[derive(Debug)]
+pub struct FeQueryMessage {
+    pub body: Bytes,
+}
+
+// We only support the simple case of Parse on unnamed prepared statement and
+// no params
+#[derive(Debug)]
+pub struct FeParseMessage {
+    pub query_string: Bytes,
+}
+
+#[derive(Debug)]
+pub struct FeDescribeMessage {
+    pub kind: u8, // 'S' to describe a prepared statement; or 'P' to describe a portal.
+                  // we only support unnamed prepared stmt or portal
+}
+
+// we only support unnamed prepared stmt and portal
+#[derive(Debug)]
+pub struct FeBindMessage {}
+
+// we only support unnamed prepared stmt or portal
+#[derive(Debug)]
+pub struct FeExecuteMessage {
+    /// max # of rows
+    pub maxrows: i32,
+}
+
+// we only support unnamed prepared stmt and portal
+#[derive(Debug)]
+pub struct FeCloseMessage {}
+
+impl FeMessage {
+    // read one message from the stream
+    pub fn read(stream: &mut impl Read) -> anyhow::Result<Option<FeMessage>> {
+        // Each libpq message begins with a message type byte, followed by message length
+        // If the client closes the connection, return None. But if the client closes the
+        // connection in the middle of a message, we will return an error.
+        let tag = match stream.read_u8() {
+            Ok(b) => b,
+            Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => return Ok(None),
+            Err(e) => return Err(e.into()),
+        };
+        let len = stream.read_u32::<BE>()?;
+
+        // The message length includes itself, so it better be at least 4
+        if len < 4 {
+            bail!("invalid message length: parsing u32");
+        }
+        let bodylen = len - 4;
```

Comment:
Not your code, but would `checked_sub` be cleaner here

Reply:
done